arangod/Agency/NodeLoadInspector.h

```diff
@@ -0,0 +1,311 @@
+      typename Base::template QualifiedVariant<Ts...>& variant) -> Status {
+    if (auto res = loadTypeField(variant.typeField, type); !res.ok()) {
+      return res;
+    }
+
+    auto node = _node->get(variant.valueField);
+    if (not node.has_value()) {
+      return {"Variant value field \"" + std::string(variant.valueField) +
+              "\" is missing"};
+    }
+    data = &node->get();
+    return {};
+  }
+
+  Status loadTypeField(std::string_view fieldName, std::string_view& result) {
+    auto v = _node->get(fieldName);
+    if (not v.has_value()) {
+      return {"Variant type field \"" + std::string(fieldName) +
+              "\" is missing"};
+    }
+    auto val = v->get().getStringView();
+    if (not val.has_value()) {
+      return {"Variant type field \"" + std::string(fieldName) +
+              "\" must be a string"};
+    }
+    result = *val;
+    return {};
+  }
+
+  template<class T>
+  bool shouldTryType(velocypack::ValueType type) {
+    using velocypack::ValueType;
+    auto isInt = [](ValueType type) {
+      return type == ValueType::Int || type == ValueType::UInt ||
+             type == ValueType::SmallInt;
+    };
+    // we try to rule out some cases where we know that the parse attempt will
+    // definitely fail, but if none of those match, we always need to simply
+    // try to parse the value
+    if constexpr (std::is_same_v<T, std::string> ||
+                  std::is_same_v<T, std::string_view>) {
+      return type == ValueType::String;
+    } else if constexpr (std::is_same_v<T, bool>) {
+      return type == ValueType::Bool;
+    } else if constexpr (std::is_integral_v<T>) {
+      return isInt(type);
+    } else if constexpr (std::is_floating_point_v<T>) {
+      return type == ValueType::Double || isInt(type);
+    } else if constexpr (detail::IsTuple<T>::value) {
+      return type == ValueType::Array && _node->isArray() &&
+             detail::TupleSize<T>::value == _node->slice().length();
+    } else if constexpr (detail::IsListLike<T>::value) {
+      return type == ValueType::Array;
+    } else if constexpr (detail::IsMapLike<T>::value) {
+      return type == ValueType::Object;
+    }
+    return true;
+  }
+
+  template<std::size_t Idx, std::size_t End, class T>
+  [[nodiscard]] Status processTuple(T& data) {
+    auto slice = _node->getArray();
+    assert(slice != nullptr);
+    if constexpr (Idx < End) {
+      auto ff = make(slice->operator[](Idx));
+      if (auto res = process(ff, std::get<Idx>(data)); !res.ok()) {
+        return {std::move(res), std::to_string(Idx), Status::ArrayTag{}};
+      }
+      return {processTuple<Idx + 1, End>(data)};
+    } else {
+      return {};
+    }
+  }
+
+  Status checkArrayLength(std::size_t arrayLength) {
+    auto slice = _node->getArray();
+    assert(slice != nullptr);
+    if (slice->length() != arrayLength) {
+      return {"Expected array of length " + std::to_string(arrayLength)};
```

Comment:
StrCat

Reply:
Why would I want to introduce a dependency on abseil just for creating an error message?

Reply:
Why introduce?
We already depends, if sometime for some reason we decide we don't want replace StrCat by any custom function is easy.

Reply:
No, this code does not. For example, I locally have a small test executable that only contains the inspector tests, and that executable currently has no dependency on abseil, and I see absolutely no reason to add one.

What exactly would be the advantage of using StrCat here? This is a very unlikely error case that definitely does not warrant any kind of optimization (performance or otherwise).

Reply:
> No, this code does not.

It is agency, not lib inspection, agency depends on arangodb library, which depends on abseil (because of few reasons s2, hashtables, strings, mutex)

So why doesn't?

> What exactly would be the advantage of using StrCat here?

It simpler (don't need to_string, which can allocate with libstdc++ (sso size 15)).

Also if we always will concat strings via abseil, it will be unified in all our code, smaller chance to write inefficient code in not error path

Reply:
Yes, this file currently lives in Agency instead of lib/inspection, but I am not sure if it will stay there. Nevertheless, this particular header file currently has no dependency on abseil, and since this is an unlikely error case I see not reason for "efficiency optimizations". Regarding simplicity - syntactically we trade one function call for another, so I don't see the big gain here.

Reply:
> but I am not sure if it will stay there

Maybe todo in such case?
About moving it out

About discussion at all.
I think in agency, and other arangod parts, abseil dependency completely ok, because we have more largest stuff, like v8, rocksdb, etc.
So build time for absl::strings is not the issue.

At the same time, code looks at least the same or better (because you don't need to explicit call for each numeric argument). And works also better (I agree it's commonly not important, but it's still true) 

So I think for implementation, not lib code, using abseil almost always better than manual concat.

Anyway fill free to resolve it, if I didn't convince you or you thinking about move this code out of agency